### PR TITLE
fix(deployment): duplicated shadow deployment if restart required

### PR DESCRIPTION
**Pull request by tilo-chen**
**Issue #, if available:**

**Description of changes:**
Added one more check when processing shadow update.

**Why is this change necessary:**
Currently, if a shadow deployment requires restart (aka kernel update workflow), ShadowDeploymentListener will create the same deployment again after restart.

Before the restart, the deployment will have been reported as IN_PROGRESS, which is a non-terminal state. Because it could be due to power off or other unexpected failure, we decided to allow it to create the deployment if the reported status was non-terminal (IN_PROGRESS). But if we're restarting normally, we're still adding the deployment which is duplicated.

If we persist the lastConfigurationArn through restart, then this will be caught by existing logic and won't create duplicated deployment. However, it will also bypass the existing logic to catch unexpected restart. Therefore, we're adding new logic to handle the above scenario.

**How was this change tested:**

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests
 - [ ] If your code makes a remote network call, it was tested with a proxy

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 1]
```diff
@@ -308,6 +314,24 @@ protected void shadowUpdated(Map<String, Object> desired, Map<String, Object> re
...
+                                ((DeploymentService) deploymentServiceLocateResult).getCurrentDeploymentTaskMetadata();
+                        if (currentDeployment != null && currentDeployment.getDeploymentId().equals(configurationArn)) {
+                            logger.atInfo().kv(CONFIGURATION_ARN_LOG_KEY_NAME, configurationArn)
+                                    .log("Ongoing deployment. Ignoring shadow update at startup");
+                            return;
+                        }
+                    }
+                } catch (ServiceLoadException e) {
+                    logger.atError().setCause(e).log("Failed to find deployment service");
+                }
```
Comment (fahadmohammed01) [inline src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:334]: we have a dedup logic on line 336, may be update that to store lastConfigurationArn across kernel restarts by storing it as a topic?

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 2]
```diff
@@ -291,9 +299,10 @@ protected void shadowUpdated(Map<String, Object> desired, Map<String, Object> re
         }
         boolean cancelDeployment = DESIRED_STATUS_CANCELED.equals(desiredStatus);
         synchronized (ShadowDeploymentListener.class) {
+            String lastConfigurationArn = (String) lastConfigurationArnTopic.getOnce();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:302]: use `Coerce.toString(lastConfigurationArnTopic)`

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 3]
```diff
@@ -67,6 +70,9 @@
     public static final String GGC_VERSION_KEY = "ggcVersion";
     public static final String DESIRED_STATUS_CANCELED = "CANCELED";
     public static final String DEPLOYMENT_SHADOW_NAME = "AWSManagedGreengrassV2Deployment";
+    public static final String SHADOW_LAST_CONFIG_ARN_TOPIC_KEY = "ShadowLastConfigurationArn";
+    @Inject
+    private Kernel kernel;
```
[the remark this answers is not in the document]
Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:75]: Trying to get CurrentDeploymentTaskMetadata here. AFAIK here we'll have to use kernel to locate DeploymentService to get it...

[the remark this answers is not in the document]
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:75]: Look at the diff... you don't user kernel at all except for `getConfig`

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 4]
```diff
@@ -308,6 +314,24 @@ protected void shadowUpdated(Map<String, Object> desired, Map<String, Object> re
                             .log("Deployment result already reported. Ignoring shadow update at startup");
                     return;
                 }
+                // Ignore if it's the ongoing deployment. This can happen if the last shadow deployment caused restart
+                try {
+                    // Using locate instead of injection here because DeploymentService lacks usable injection
+                    // constructor. Same as in IotJobsHelper.evaluateCancellationAndCancelDeploymentIfNeeded
+                    GreengrassService deploymentServiceLocateResult =
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:321]: you don't need to do this, you should be able to just `@inject` the deployment service itself.

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 5]
```diff
@@ -68,6 +72,8 @@
     public static final String DESIRED_STATUS_CANCELED = "CANCELED";
     public static final String DEPLOYMENT_SHADOW_NAME = "AWSManagedGreengrassV2Deployment";
     @Inject
+    private Kernel kernel;
+    @Inject
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:76]: why is this added? How was it working before?

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java (hunk 4 above)
Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:321]: Sadly, directly `@inject DeploymentService` gives no usable injection constructor error

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java (hunk 1 above)
Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:334]: Persisting it across restart can avoid the duplicated deployment. But it also prevents us from retrying deployment that failed because of power cycle in the middle of a deployment.
It's still necessary to check the ongoing deployment ID to tell which case we're dealing with.

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java (hunk 5 above)
Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:76]: this is just to get `DeploymentService.getCurrentDeploymentTaskMetadata`. We didn't need that info. But in order to tell why last deployment didn't finish, (unexpected failure vs. managed restart) we can check the ongoing deployment ID. If it was a managed restart, the ongoing deployment should be the same.

File: src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java [hunk 6]
```diff
@@ -308,6 +314,24 @@ protected void shadowUpdated(Map<String, Object> desired, Map<String, Object> re
...
+                // Ignore if it's the ongoing deployment. This can happen if the last shadow deployment caused restart
+                try {
+                    // Using locate instead of injection here because DeploymentService lacks usable injection
+                    // constructor. Same as in IotJobsHelper.evaluateCancellationAndCancelDeploymentIfNeeded
+                    GreengrassService deploymentServiceLocateResult =
+                            kernel.locate(DeploymentService.DEPLOYMENT_SERVICE_TOPICS);
+                    if (deploymentServiceLocateResult instanceof DeploymentService) {
+                        DeploymentTaskMetadata currentDeployment =
+                                ((DeploymentService) deploymentServiceLocateResult).getCurrentDeploymentTaskMetadata();
+                        if (currentDeployment != null && currentDeployment.getDeploymentId().equals(configurationArn)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/ShadowDeploymentListener.java:326]: swap the `.equals` check to prevent possible NPE

**Review by MikeDombo — DISMISSED** [review]

**Review by fahadmohammed01 — DISMISSED** [review]

**Review by MikeDombo — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]

**Review by abanthiy — APPROVED** [review]
